### ISSUE TITLE
RESTEasy Reactive: Reactive filter hangs the request when reading the body

### ISSUE THREAD

**Issue by csotiriou** [quarkusio/quarkus#23263]
### Describe the bug

When using RESTEasy Reactive with a Reactive filter, if we try to read the request body inside the filter, the filter is not terminating at all, so the request never returns a response.

Note that this happens even when paying attention to making the stream read in the Quarkus worker thread, and then returning back to the vert.x thread, so it should not have problems there.

### Expected behavior

When using a request filter and reading the body inside the filter, the filter should complete properly and the HTTP request should proceed and be successful.

### Actual behavior

The http request never returns a response when the filter is used.

### How to Reproduce?

1. Download the example project
2. Run the project using `mvn compile quarkus:dev`
3. Make a request using the following curl:

```
curl --request POST \
--url http://localhost:8080/getDummyId \
--header 'Accept: application/json' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data id=1
```

You will notice that the request never completes.

### Output of `uname -a` or `ver`

Darwin 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64

### Output of `java -version`

OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

### Quarkus version or git rev

2.6.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven

### Additional information

Note that after reading the HTTP Request body, we pay attention to setting a new InputStream to the http request using `setEntityStream`. This is a methodology that has worked nicely in the non-reactive version of RESTEasy. So one would expect that the filter would not have a problem doing the same thing using the same APIs in RESTEasy Reactive.

**Comment by geoand** [comment]
Juggling threads like that isn't really supported to be honest.

The point of returning `Uni` from a server request filter, is to allow RESTEasy Reactive is to not block the current thread serving the request. RESTEasy Reactive is the one that will decide where to proceed when the `Uni` is complete. 

@cescoffier I don't support we have a way of printing a warning (or even failing) when `emitOn` is called, do we?

**Comment by cescoffier** [comment]
@geoand no, we don't. `emitOn` is a valid operator that sometimes makes sense. That being said, when used in such a context, you need to go back to the I/O thread. Once you are done with the part of the code, that must be executed on another thread.

I gave an example of such a code in https://stackoverflow.com/questions/70802959/bring-uni-event-back-to-the-caller-thread.

**Comment by csotiriou** [comment]
@cescoffier what you have mentioned is indeed what we are doing here in the attached sample code.

Based on this, then is this issue still valid @cescoffier ?

**Comment by cescoffier** [comment]
I would need to check. Is the body handler already set at that point (@geoand)?

**Comment by geoand** [comment]
It is not ready, no

**Comment by cescoffier** [comment]
So, that's the issue. You cannot wait for the body to be read if there is no body handler yet.

**Comment by geoand** [comment]
Indeed. This would require https://github.com/quarkusio/quarkus/issues/17280

**Comment by csotiriou** [comment]
Thank you for your answers, I'm not quite sure there is a definitive solution given in #17280, unless I miss something

From what I see in #17280 there are some possible recommendations:
- use @stuartwdouglas' recommendation of injecting `org.jboss.resteasy.reactive.server.spi.ServerHttpRequest` and read the body. I would assume that this is asynchronous, so if we put this inside the filter, the callback will be called possible after the filter ends (am I right?)
- use the `request.body().` method (but would that consume the body without leaving it available for the main request?)
- Apply the logging filter afterward in the filter chain so that we can give it the opportunity to be read (would that be a viable scenario?)

**Comment by IsSkyfalls** [comment]
My use case for this is to calculate a hmac of the whole request body. Currently, there are no ways of reading the body in a non-blocking filter. Even if you use `@ServerRequestFilter(readBody = true)`, `request.body()` still doesn't complete, until all the filters has returned(I can't pinpoint exactly when). So if you return a Uni that waits for `Future<Buffer>`, the filter will be stuck. Reading the InputStream directly blocks the io thread and is not allowed.

**Comment by silviu-negoita** [comment]
Hello, any updates on this? I need to read the request body for logging purpose. ATM, I am stuck with `org.jboss.resteasy.reactive.common.core.BlockingNotAllowedException: Attempting a blocking read on io thread` when trying `containerRequestContext.getEntityStream().readAllBytes()`

**Comment by csotiriou** [comment]
@silviu-negoita using request.body() seems to be working

```
        requestContext.body().toCompletionStage().thenAccept(buffer -> {
            log.info("requestBody in filter: " + buffer.toString());
        });
```

However, I would like to ask if this is a viable solution, @geoand ? Does using this solution interfere with the Quarkus body readers? I have also tested @stuartwdouglas' proposal (using `org.jboss.resteasy.reactive.server.spi.ServerHttpRequest`, and it doesn't work since Quarkus uses the body listener internally, so it overwrites everything we put there.

**Comment by geoand** [comment]
@csotiriou where are you using that?

**Comment by csotiriou** [comment]
I am using the `requestContext.body()` in a `@ServerRequestFilter()`, for logging purposes. I tried also injecting `org.jboss.resteasy.reactive.server.spi.ServerHttpRequest` there using `var request = CurrentRequestManager.get().serverRequest();` but it's doesn't work as Quarkus also uses it internally and resets it, so I don't believe we should override that.

**Comment by csotiriou** [comment]
Well, I should point out the following however,

the `request.body()` is supposed to be fulfilled **after** the filter. If you try to do something like this, it won't work:

```java
@ServerRequestFilter
    public Uni<RestResponse<?>> filter(HttpServerRequest request) {
        return Uni.createFrom().completionStage(request.body().toCompletionStage())
                .map(buffer -> {
                    System.out.println(buffer.toString());
                    return buffer;
                })
                .map(buffer -> null);
    }
```

Because the `body()`method will be fulfilled **after** the filter, but in this case the filter will finish after the body has read (deadlock).

That effectively deters anyone from doing authorization checks in filters where the body is needed to be read before we proceed.

**Comment by csotiriou** [comment]
This discussion thread seems to be interesting:
https://github.com/quarkusio/quarkus/pull/29825 - Could the new annotation `@WithFormRead` be

**Comment by geoand** [comment]
Very likely

**Comment by csotiriou** [comment]
It seems that currently this supports only `application/x-www-form-urlencoded` . Can a possible workaround be this?

```java
@Provider
public class MyFilter {

    @ServerRequestFilter()
    @Blocking
    public Uni<Void> filter(ResteasyReactiveContainerRequestContext context) throws IOException {
        //print the current thread
        System.out.println("Thread = " + Thread.currentThread().getName());
        return Vertx.currentContext().executeBlocking(Uni.createFrom().item(() -> {
            //print the current thread
            System.out.println("Thread = " + Thread.currentThread().getName());
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            try {
                IOUtils.copy(context.getEntityStream(), baos);
            } catch (IOException e) {
                return Uni.createFrom().failure(e);
            }
            byte[] bytes = baos.toByteArray();
            System.out.println("bytes = " + new String(bytes));
            context.setEntityStream(new ByteArrayInputStream(bytes));
            return new String(bytes);
        })).replaceWithVoid();
    }
}
```

The following dependencies must be added

```
        <dependency>
            <groupId>commons-io</groupId>
            <artifactId>commons-io</artifactId>
            <version>2.11.0</version>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-vertx</artifactId>
        </dependency>
```

My only worry would be that in order to continue the filter chain correctly, I am setting a new ByteArrayInputStream, and not the default Vertx Stream that is originally present in the context. Would this cause problems?

**Comment by ichrist-gr** [comment]
Well, i guess a different approach would be to perform your authorization chekcs with a CDI interceptor and invoke it before your jax-rs endpoint. With interceptor the body will be already registered from the container so you can extract it from your http method parameters.

**Comment by jnbdz** [comment]
The hang is gone on current main, but reading the body in a filter still depends on which thread the filter runs on, and for a non-form body on the event loop there is no mechanism at all. Measured with a `Uni` returning `@ServerRequestFilter` in the rest deployment module:

| case | thread | result |
| --- | --- | --- |
| JSON body, plain filter, endpoint blocking (the request from the description) | `executor-thread-1` | body is read, request proceeds |
| JSON body, filter annotated `@WithFormRead`, endpoint blocking | worker | body is read, request proceeds |
| form body, `@WithFormRead`, values read from the parsed form | worker | values are visible |
| form body, `@WithFormRead`, raw `getEntityStream()` | worker | `IOException: Stream is closed` |
| JSON body, plain filter, endpoint `@NonBlocking` | `vert.x-eventloop-thread-1` | `BlockingNotAllowedException: Attempting a blocking read on io thread` |
| JSON body, `@WithFormRead`, endpoint `@NonBlocking` | `vert.x-eventloop-thread-1` | `BlockingNotAllowedException` |
| `@Blocking` on the filter method itself | n/a | build fails: the annotation may only be used on entrypoint methods |

The fourth row is expected: `@WithFormRead` has the form read and parsed before the filters run, so the stream is consumed by then and the values come from the parsed form.

The last three rows are the remaining gap. `@WithFormRead` does not help for a body that is not a form: `FormBodyHandler` only replaces the entity stream with a buffered copy when the endpoint declares a body parameter, so a JSON request whose endpoint takes no body parameter never gets a readable stream, and the read stays a blocking read on the event loop. This is what the hmac and request logging use cases in this thread need, and there is currently no supported way to do it when the filter runs on the event loop.

Would you like a mechanism for this, something like `@WithFormRead` generalised to any content type (pre-read the body and hand the filter a buffered stream), either by widening that annotation or adding a separate one? I am happy to open the PR, I would just rather agree on the shape first.

**Comment by cescoffier** [comment]
@geoand what do you think?

**Comment by geoand** [comment]
I plan to have a look soon

**Comment by geoand** [comment]
> Would you like a mechanism for this, something like @WithFormRead generalised to any content type (pre-read the body and hand the filter a buffered stream), either by widening that annotation or adding a separate one? I am happy to open the PR, I would just rather agree on the shape first.


Unfortunately I don't understand what is being proposed here. Can you please elaborate?
